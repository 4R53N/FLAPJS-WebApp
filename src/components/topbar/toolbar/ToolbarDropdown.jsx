import React from 'react';
import PropTypes from 'prop-types';
// import Style from './ToolbarDropdown.module.css';

/**
 * A React component that can do anything you want. :D.
 *
 * @param e
 */
class ToolbarDropdown extends React.Component 
{
    constructor(props) 
    {
        super(props);
        this.state = {
            listOpen: false,
        };
        this.handleClick = this.handleClick.bind(this);
    }

    /** @override */
    componentDidMount() 
    {
        document.addEventListener('mousedown', this.handleClick, false);
    }
    
    /** @override */
    componentWillUnmount() 
    {
        document.removeEventListener('mousedown', this.handleClick, false);
    }

    handleClick(e)
    {
        if (this.node.contains(e.target)) 
        {
            return;
        }
        this.handleClickOutside();
    }

    handleClickOutside() 
    {
        this.setState({
            listOpen: false
        });
    }

    toggleList() 
    {
        this.setState(prevState => ({
            listOpen: !prevState.listOpen
        }));
    }

    /** @override */
    render()
    {
        const props = this.props;
        const{ listOpen } = this.state;

        return (
            <div
                id={props.id}
                style={props.style}
                ref = {node => this.node = node}
                className={props.className}
            >
                <button className="dd-header" onClick={() => this.toggleList()}>
                    <div className="dd-header-title">{props.title}</div>
                    {/* {listOpen ? <p>up</p> : <p>down</p>} */}
                </button>
                {listOpen && 
                <ul className="dd-list">
                    {props.list.map((item) => (
                        <li key={item.id}>
                            <button
                                className="dd-list-item"
                                onClick={() => {props.toggleItem(item.id, item.key);}}>
                                {item.title}
                                {item.selected && <p>x</p>}
                            </button>
                        </li>
                    ))}
                </ul>}
            </div>
        );
    }
}

ToolbarDropdown.propTypes = {
    id: PropTypes.string,
    style: PropTypes.object,
    className: PropTypes.string,
    title: PropTypes.string,
    list: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number,
        title: PropTypes.string,
        selected: PropTypes.bool,
        key: PropTypes.string
    })),
    toggleItem: PropTypes.func
};
ToolbarDropdown.defaultProps = {
};

export default ToolbarDropdown;
