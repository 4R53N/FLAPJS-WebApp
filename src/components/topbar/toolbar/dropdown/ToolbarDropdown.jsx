import React from 'react';
import PropTypes from 'prop-types';
import Style from './ToolbarDropdown.module.css';

import ToolbarDropdownOptions from '@flapjs/components/topbar/toolbar/dropdown/ToolbarDropdownOptions.jsx';

/**
 * A React component that can do anything you want. :D.
 *
 * @param e
 */
class ToolbarDropdown extends React.Component 
{
    constructor(props) 
    {
        super(props);
        this.state = {
            listOpen: false,
        };
        this.node = React.createRef();
        this.handleClick = this.handleClick.bind(this);
    }
    
    /** @override */
    componentDidMount() 
    {
        document.addEventListener('mousedown', this.handleClick, false);
    }
    
    /** @override */
    componentWillUnmount() 
    {
        document.removeEventListener('mousedown', this.handleClick, false);
    }
    
    handleClick(e)
    {
        if ((this.node.current.contains(e.target) && !this.state.listOpen)) 
        {
            this.setState({
                listOpen: true
            });
        }
        else 
        {
            this.setState({
                listOpen: false
            });
        }
    }

    /** @override */
    render()
    {
        const props = this.props;
        const{ listOpen } = this.state;

        return (
            <div
                id={props.id}
                style={props.style}
                ref = {this.node}
                className={Style.container}
            >
                {/* <button className="dd-header" onClick={e => this.handleClick(e)}> */}
                <button>
                    <div>{props.title}</div>
                    {/* {listOpen ? <p>up</p> : <p>down</p>} */}
                </button>
                {listOpen && 
                // {true && 
                <ul className={Style.dropdownlist}>
                    {props.list.map((item) => (
                        <li key={item.id}>
                            {
                                item.options ?
                                    <ToolbarDropdownOptions
                                        title={item.title}
                                        list={item.options}
                                        toggleItem={props.toggleItem}
                                    />
                                    :
                                    <button
                                        onMouseDown={() => {props.toggleItem(item.id, item.key);}}>
                                        {item.title}
                                        {item.selected && <p>x</p>}
                                    </button>
                            }
                        </li>
                    ))}
                </ul>}
            </div>
        );
    }
}

ToolbarDropdown.propTypes = {
    id: PropTypes.string,
    style: PropTypes.object,
    className: PropTypes.string,
    title: PropTypes.string,
    list: PropTypes.array,
    toggleItem: PropTypes.func
};
ToolbarDropdown.defaultProps = {
};

export default ToolbarDropdown;
