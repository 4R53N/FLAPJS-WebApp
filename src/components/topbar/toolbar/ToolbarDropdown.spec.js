import React from 'react';
import { shallow } from 'enzyme';

import ToolbarDropdown from './ToolbarDropdown.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(<ToolbarDropdown/>);
});

describe('<ToolbarDropdown/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<ToolbarDropdown/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<ToolbarDropdown/> lifecycle', () =>
{
    /** ...later. */
});
