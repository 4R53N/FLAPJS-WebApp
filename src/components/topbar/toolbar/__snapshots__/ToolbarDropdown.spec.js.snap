// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToolbarDropdown/> rendering renders correctly 1`] = `
<div>
  <button
    className="dd-header"
    onClick={[Function]}
  >
    <div
      className="dd-header-title"
    />
  </button>
</div>
`;
